`Communication`: Allow user to forward messages

asliayk opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check all tasks by putting an x in the [ ] (don't: [x ], [ x], do: [x]). Remove not applicable tasks and do not leave them unchecked -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- Remove tasks that are not applicable for your PR. Please only put the PR into ready for review, if all relevant tasks are checked! -->
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable, and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [X] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [X] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).


#### Server
- [X] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [X] I **strictly** followed the principle of **data economy** for all database calls.
- [X] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [X] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [X] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).


#### Client
- [X] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [X] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [X] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [X] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [X] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [X] I added multiple screenshots/screencasts of my UI changes.
- [X] I translated all newly inserted strings into English and German.


### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Currently user cannot forward messages to other channels/direct-group chats.

(Closes #9066)


### Description
<!-- Describe your changes in detail -->
The forward message feature has been added. Users can now navigate to the forward message dialog by clicking on the forward message option. They can select a random number of destinations and add content to the forwarded message (similar to Slack).

`forwarded_message` table has been added to the database with the following fields:

- **id**: id of forwarded message
- **source_id**: the id of the original message
- **source_type**: the type of the original message (post or answer)
- **destination_post_id**: the id of the destination post (the id of newly created post with the forwarded message)
- **destination_answer_id**: the id of the destination answer post (the id of newly created answer post with the forwarded message)

`post` and `answer_post` table have been updated with a new column named **has_forwarded_messages**.


Known Issue for a followup PR:
The server-side changes for forwarding multiple messages within a single message and forwarding a message as a reply to another message (inside an AnswerPost) are already in place, but the client-side implementation is not ready yet. To achieve this, a unique message link needs to be generated for each post/answer post, similar to the "Copy Link" feature in Slack. This will be completed in a follow-up PR.


### Steps for Testing
<!-- Please describe in detail how reviewers can test your changes. Make sure to take all related features and views into account! Below is an example that you can refine. -->
Prerequisites:
- 1 Instructor/Student
- 1 Course with Communication enabled

1. Log in to Artemis.
2. Navigate to the Communication section of a course.
3. Select a random message to forward or create a new one. Click on the forward message option (available on its hover bar or in the dropdown menu when opened via right-click).
4. Add some content to the forwarded message and select random destinations to forward the message.
5. Go to the destination channels/user chats and observe the forwarded messages.



### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked
> Click on the badges to get to the test servers.

[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)](https://artemis-test1.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)](https://artemis-test2.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)](https://artemis-test3.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)](https://artemis-test4.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)](https://artemis-test5.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)](https://artemis-test6.artemis.cit.tum.de)
[![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)](https://artemis-test9.artemis.cit.tum.de)

### Review Progress
<!-- Each PR should be reviewed by at least two other developers. The code, the functionality (= manual test) and the exam mode need to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->
<!-- All PRs that might affect the exam mode (e.g. change a client component that is also used in the exam mode) need an additional verification that the exam mode still works. -->

#### Performance Review
<!-- See [Large Course Setup](https://github.com/ls1intum/Artemis/tree/develop/supporting_scripts/course-scripts/quick-course-setup) for the automation script that handles large course setup. -->
- [X] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance even for very large courses with more than 2000 students.
- [X] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [X] Code Review 1
- [X] Code Review 2
#### Manual Tests
- [X] Test 1
- [X] Test 2
#### Performance Tests
- [X] Test 1
- [X] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- The line coverage must be above 90% for changes files and you must use extensive and useful assertions for server tests and expect statements for client tests. -->
<!-- Note: Use the table below and confirm in the last column that you have implemented extensive assertions for server tests and expect statements for client tests. -->
<!--       You can use `supporting_script/generate_code_cov_table/generate_code_cov_table.py` to automatically generate one from the corresponding Bamboo build plan artefacts. -->
<!--       Remove rows with only trivial changes from the table. -->
<!--
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ExerciseService.java | 85% | ✅                           |
| programming-exercise.component.ts | 95% | ✅              |
-->
#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| answer-post.model.ts | 100% | ✅ |
| forwarded-message.model.ts | 91.66% | ✅ ❌ |
| post.model.ts | 100% | ✅ |
| posting.model.ts | 86.36% | ✅ ❌ |
| course-conversations.component.ts | 91.3% | ✅ ❌ |
| forward-message-dialog.component.ts | 95.79% | ✅ ❌ |
| conversation-messages.component.ts | 95.53% | ✅ ❌ |
| answer-post.service.ts | 100% | ✅ |
| answer-post.component.ts | 96.29% | ✅ ❌ |
| forwarded-message.service.ts | 100% | ✅ |
| forwarded-message.component.ts | 90% | ✅ ❌ |
| metis-conversation.service.ts | 83.47% | ✅ ❌ |
| metis.service.ts | 88.47% | ✅ ❌ |
| post.service.ts | 80% | ✅ ❌ |
| post.component.ts | 97.08% | ✅ ❌ |
| posting-reactions-bar.component.ts | 83.67% | ✅ ❌ |
| posting-thread.component.ts | 100% | ✅ |
| posting.directive.ts | 93.82% | ✅ ❌ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| AnswerPost.java | 85% | ✅ ❌ |
| ForwardedMessage.java | 42% | ✅ ❌ |
| Post.java | 91% | ✅ ❌ |
| ForwardedMessageDTO.java | 100% | ✅ |
| ForwardedMessagesGroupDTO.java | 100% | ✅ |
| ForwardedMessageRepository.java | not found (added) | ❌ |
| AnswerMessageService.java | 88% | ✅ ❌ |
| ConversationMessagingService.java | 92% | ✅ ❌ |
| AnswerMessageResource.java | 90% | ✅ ❌ |
| ConversationMessageResource.java | 88% | ✅ ❌ |
| ForwardedMessageResource.java | 94% | ✅ ❌ |



### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. Remove the section if you did not change the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->

**forwarded message view**
![image](https://github.com/user-attachments/assets/f2416c55-edc8-45af-bc1b-0bd89e66b641)

**forward message dialog**
![image](https://github.com/user-attachments/assets/defe694c-d37c-43e2-81cc-1fe418094e35)



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced robust message forwarding, allowing users to forward posts and replies to selected recipients.
  - Launched an intuitive forward message dialog featuring autocomplete for recipient selection, rich text editing, and preview capabilities.
  - Enhanced conversation navigation by clearly indicating forwarded content, ensuring users can easily trace back to original posts.
  - Updated user interface elements such as icons and tooltips to streamline message actions and improve overall messaging experience.
  - Added functionality to retrieve and manage forwarded messages, enhancing user interaction with posts.
  - Implemented new methods for batch retrieval of messages and improved data handling for forwarded content.
  - Enhanced the `Post` and `AnswerPost` classes to manage forwarded messages effectively.
  - Introduced a new `ForwardedMessage` entity and related DTOs for improved message management.
  - Added support for retrieving forwarded messages based on destination IDs.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/domain/ForwardedMessage.java:
```diff
@@ -0,0 +1,119 @@
+
+    @NotNull
+    @Column(name = "source_id", nullable = false)
+    private Long sourceId;
```
can you use `long` instead? Then it can never be null

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/domain/ForwardedMessage.java:
```diff
@@ -0,0 +1,119 @@
+@Entity
+@Table(name = "forwarded_message")
+@Check(constraints = "((destination_post_id IS NOT NULL AND destination_answer_id IS NULL) OR (destination_post_id IS NULL AND destination_answer_id IS NOT NULL))")
+public class ForwardedMessage {
```
please inherit from `DomainObject`

krusche commented on src/main/resources/config/liquibase/changelog/20241129213030_changelog.xml:
```diff
@@ -0,0 +1,62 @@
+
+        <!-- Add check constraint for destination columns -->
+        <sql>
+            ALTER TABLE forwarded_message
```
please make sure this works for mysql and postgres

krusche commented on src/main/webapp/app/entities/metis/forwarded-message.model.ts:
```diff
@@ -0,0 +1,31 @@
+    public id?: number;
+    public sourceId?: number;
+    public sourceType?: PostingType;
+    public destinationPost: Post | undefined;
```
```suggestion
    public destinationPost?: Post;
```

krusche commented on src/main/webapp/app/entities/metis/forwarded-message.model.ts:
```diff
@@ -0,0 +1,31 @@
+    public sourceId?: number;
+    public sourceType?: PostingType;
+    public destinationPost: Post | undefined;
+    public destinationAnswerPost: AnswerPost | undefined;
```
```suggestion
    public destinationAnswerPost?: AnswerPost;
```

krusche commented on src/main/webapp/app/entities/metis/forwarded-message.model.ts:
```diff
@@ -0,0 +1,31 @@
+    public sourceType?: PostingType;
+    public destinationPost: Post | undefined;
+    public destinationAnswerPost: AnswerPost | undefined;
+    public content: string | undefined;
```
```suggestion
    public content: string?;
```

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ForwardedMessageResource.java:
```diff
@@ -0,0 +1,102 @@
+     *         and the value is a set of forwarded message DTOs grouped by the destination post IDs
+     */
+    @GetMapping("/forwarded-messages/posts")
+    public ResponseEntity<Map<Long, Set<ForwardedMessageDTO>>> getForwardedMessagesByDestPostIds(@RequestParam Set<Long> dest_post_ids) {
```
in Java we do not use "_" --> `destPostIds`

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ForwardedMessageResource.java:
```diff
@@ -0,0 +1,102 @@
+     * @return the ResponseEntity with status 200 (OK) and with body containing a map where the key is the destination post ID
+     *         and the value is a set of forwarded message DTOs grouped by the destination post IDs
+     */
+    @GetMapping("/forwarded-messages/posts")
```
This URL is not REST conform

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ForwardedMessageResource.java:
```diff
@@ -0,0 +1,102 @@
+     * @return the ResponseEntity with status 200 (OK) and with body containing a map where the key is the destination answer post ID
+     *         and the value is a set of forwarded message DTOs grouped by the destination answer post IDs
+     */
+    @GetMapping("/forwarded-messages/answers")
+    public ResponseEntity<Map<Long, Set<ForwardedMessageDTO>>> getForwardedMessagesByDestAnswerPostIds(@RequestParam Set<Long> dest_answer_post_ids) {
```
same feedback as above

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ForwardedMessageResource.java:
```diff
@@ -0,0 +1,102 @@
+    public ResponseEntity<Map<Long, Set<ForwardedMessageDTO>>> getForwardedMessagesByDestPostIds(@RequestParam Set<Long> dest_post_ids) {
+        Set<ForwardedMessage> forwardedMessages = forwardedMessageRepository.findAllByDestinationPostIds(dest_post_ids);
+
+        Map<Long, Set<ForwardedMessageDTO>> groupedDtos = forwardedMessages.stream()
```
do you really need to return a datatype like this? `Map<Long, Set<ForwardedMessageDTO>>`. This could also be a list of a DTO with `(id, Set<ForwardedMessageDTO)`

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ForwardedMessageResource.java:
```diff
@@ -0,0 +1,102 @@
+    public ResponseEntity<Map<Long, Set<ForwardedMessageDTO>>> getForwardedMessagesByDestAnswerPostIds(@RequestParam Set<Long> dest_answer_post_ids) {
+        Set<ForwardedMessage> forwardedMessages = forwardedMessageRepository.findAllByDestinationAnswerPostIds(dest_answer_post_ids);
+
+        Map<Long, Set<ForwardedMessageDTO>> groupedDtos = forwardedMessages.stream()
```
same here, we should avoid using map for response types

krusche reviewed (changes requested):
please see my inline feedback

julian-wls reviewed (commented):
Great work! Tested on TS2.
Forwarding works great but you have to reload after the message has been forwarded as a new DM in order to see the new chat in the conversation list.

eylulnc reviewed (approved):
Tested in test session on TS2 work as expected

FelberMartin reviewed (dismissed):
Tested on TS2, works as described.

I only found a small issue regarding dark theme, the text in the search bar is hardly visible
<img width="392" alt="image" src="https://github.com/user-attachments/assets/b05382bd-f8af-4e60-a9cb-ebfe926587ee">

cremertim commented on src/main/webapp/app/overview/course-conversations/dialogs/forward-message-dialog/forward-message-dialog.component.ts:
```diff
@@ -0,0 +1,252 @@
+    newPost: Post = new Post();
+    defaultActions: TextEditorAction[];
+    editorHeight = input<MarkdownEditorHeight>(MarkdownEditorHeight.INLINE);
+    @ViewChild('searchInput') searchInput!: ElementRef;
```
can we use viewChild here?

PaRangger replied:
Also we could use the viewChild signal here

anian03 reviewed (commented):
Tested during testing session, works as described. Didn't see any issues apart from the ones already mentioned

PaRangger commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ForwardedMessageResource.java:
```diff
@@ -0,0 +1,105 @@
+        long start = System.nanoTime();
+
+        if (!"post".equalsIgnoreCase(type) && !"answer".equalsIgnoreCase(type)) {
+            throw new BadRequestAlertException("Invalid type provided. Must be 'post' or 'answer'.", "ForwardedMessage", "invalidType");
```
```suggestion
            throw new BadRequestAlertException("Invalid type provided. Must be 'post' or 'answer'.", ENTITY_NAME, "invalidType");
```

PaRangger commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ForwardedMessageResource.java:
```diff
@@ -0,0 +1,105 @@
+     * @throws BadRequestAlertException if the type parameter is invalid or unsupported.
+     */
+    @GetMapping("/forwarded-messages")
+    public ResponseEntity<List<ForwardedMessagesGroupDTO>> getForwardedMessages(@RequestParam Set<Long> ids, @RequestParam String type) {
```
Would it make sense to pass a PostingType enum here instead of the strings/what coderabbit suggested?

PaRangger commented on src/main/webapp/app/overview/course-conversations/dialogs/forward-message-dialog/forward-message-dialog.component.ts:
```diff
@@ -0,0 +1,252 @@
+    img: string;
+}
+
+@Component({
```
It would be nice if we made this standalone. 😄

PaRangger commented on src/main/webapp/app/overview/course-conversations/layout/conversation-messages/conversation-messages.component.ts:
```diff
@@ -301,7 +306,80 @@ export class ConversationMessagesComponent implements OnInit, AfterViewInit, OnD
-        this.groupPosts();
+        const postIdsWithForwardedMessages = this.posts.filter((post) => post.hasForwardedMessages && post.id !== undefined).map((post) => post.id) as number[];
+
+        if (postIdsWithForwardedMessages.length > 0) {
```
Maybe we could invert some of these ifs to reduce a bit of the nesting.

asliayk replied:
I think this increases readability for now so I left it as it is

PaRangger commented on src/main/webapp/app/shared/metis/forwarded-message.service.ts:
```diff
@@ -0,0 +1,65 @@
+export class ForwardedMessageService {
+    public resourceUrl = 'api/forwarded-messages';
+
+    constructor(protected http: HttpClient) {}
```
```suggestion
    protected http = inject(HttpClient);
```

PaRangger commented on src/main/webapp/app/shared/metis/forwarded-message/forwarded-message.component.html:
```diff
@@ -0,0 +1,49 @@
+            </span>
+        </div>
+        <div class="forwarded-message-body" #messageContent [ngClass]="{ expanded: showFullForwardedMessage }">
+            <div class="preserve-line-breaks">{{ displayedForwardedContent() }}</div>
```
I think you need to use the `jhi-posting-content` component here, otherwise it wont render properly.

<img width="648" alt="Bildschirmfoto 2024-12-10 um 13 11 42" src="https://github.com/user-attachments/assets/4aab1b5f-ad4d-4974-9882-c7ee64c14021">

PaRangger commented on src/main/webapp/app/shared/metis/forwarded-message/forwarded-message.component.ts:
```diff
@@ -0,0 +1,114 @@
+    postingIsOfToday: boolean;
+    todayFlag?: string;
+    readonly onNavigateToPost = output<Posting>();
+    @ViewChild('messageContent') messageContent!: ElementRef;
```
Also here it would be nice if we used the viewChild signal

PaRangger commented on src/main/webapp/app/shared/metis/forwarded-message/forwarded-message.component.ts:
```diff
@@ -0,0 +1,114 @@
+import dayjs from 'dayjs';
+import { Conversation } from 'app/entities/metis/conversation/conversation.model';
+
+@Component({
```
Can we also make this standalone?

asliayk replied:
it caused a circular dependency (with MetisModule), so I left it as it is for now

PaRangger commented on src/main/webapp/app/shared/metis/forwarded-message/forwarded-message.component.ts:
```diff
@@ -0,0 +1,114 @@
+        this.isChannel();
+    }
+
+    ngOnChanges(changes: SimpleChanges): void {
```
Does ngOnChanges still work with input signals? I think we have to use effect({}) to react to changes here.

PaRangger commented on src/main/webapp/app/shared/metis/posting-reactions-bar/posting-reactions-bar.directive.ts:
```diff
@@ -121,6 +134,96 @@ export abstract class PostingsReactionsBarDirective<T extends Posting> implement
+
+                    modalRef.result.then(
+                        async (selection: { channels: Conversation[]; users: UserPublicInfoDTO[]; messageContent: string }) => {
+                            if (selection) {
```
Maybe its also possible here to remove some of the nesting by inverting ifs.

asliayk replied:
I chose not to refactor this for now. Although these ifs might not be necessary, I think the code seems more understandable in its current form

PaRangger commented on src/test/javascript/spec/component/overview/course-conversations/layout/conversation-messages/conversation-messages.component.spec.ts:
```diff
@@ -268,6 +270,23 @@ examples.forEach((activeConversation) => {
 
+        it('should fetch forwarded messages correctly', () => {
+            const mockForwardedMessages: ForwardedMessage[] = [
+                { id: 101, sourceId: 10, sourceType: 'POST' } as unknown as ForwardedMessage,
```
Does this not work without the double cast?

asliayk replied:
I left it this way because I needed to add more valid fields, and otherwise it would have made the code even longer.

PaRangger commented on src/test/javascript/spec/service/metis/forwarded-message.service.spec.ts:
```diff
@@ -0,0 +1,90 @@
+        expect(service).toBeTruthy();
+    });
+
+    describe('#createForwardedMessage', () => {
```
I'm not sure if we have a strict naming convention here, but usually we don't use hashtags at the beginning.

Malekos74 reviewed (approved):
### Tested on server: 02
Works great
However when writing the channel/user name, the text is black for some reason
![image](https://github.com/user-attachments/assets/fc4d8307-32a2-452b-b625-d4f59897e773)

eylulnc reviewed (approved):
Tested in TS2, works as expected.
Forward is reachable via hoover. ✅
Message is forwarded to selected channel ✅

FelberMartin reviewed (dismissed):
Tested on TS2, works as expected. Good work!

PaRangger reviewed (approved):
Tested on TS2. Works as expected. 👍 Some small remarks: I noticed that the announcement channel gets displayed to students too, which leads to an error message (see the gif). Additionally the text is not displayed in rendered form in the "forward message" modal, which I'm unsure whether or not it is intentional.

![forward](https://github.com/user-attachments/assets/9fa7bf30-8dc0-4945-aa99-e56b6d6ed946)

julian-wls reviewed (approved):
Tested on TS2 during session. Works as expected!
Well done :)

cremertim reviewed (dismissed):
Code changes lgtm overall. Minor thing, there are a lot of relative imports in your test files. Would be good to change those to absolute paths if possible, but should not hinder the PR from being merged

anian03 reviewed (commented):
Only noticed the same thing as Paul, we are displaying the announcement channel to students, despite them not being able to post there. Would be great if you could fix that, otherwise no opposition from me either :)

asliayk commented:
thank you for your reviews :) after your reviews I updated the code to address the following issues:

- Users can no longer forward a message to an announcement channel.
- Fixed the rendering issue in the forward message dialog by using the `PostingContentComponent`.
- Resolved failing server tests by making some minor changes in `ForwardedMessage.java`.
- Added additional client tests.

I appreciate your re-reviews :)

HawKhiem reviewed (dismissed):
Tested on TS3. Works as expected

![image](https://github.com/user-attachments/assets/1183de88-3e7c-4c54-b4a6-5b05075885e6)

JerroyTan reviewed (commented):
[Tested on TS2]
Functionalities work mostly as expected, except for two quirks while I was playing around.
1. After forwarding messages, I logged in to another user and the forwarded content was not rendered at first. After refreshing, they got rendered like normal. This also happened after my 2. point.
*directly after logging in*
![image](https://github.com/user-attachments/assets/f50b1d7b-43a5-4a87-ad23-efe811f91078)
*after refreshing*
![image](https://github.com/user-attachments/assets/e5a1a5c8-80e2-4230-87f4-41f4160a725b)

2. When forwarding from a private channel to a public channel, the link for "view conversation" appears even if the user is not in the private channel, and clicking on it causes them to access a site they are not supposed to, and reloads them to the all public messages. Is this desired behaviour?
![image](https://github.com/user-attachments/assets/db8a9268-267b-4833-8950-17b792e6f808)
PS: after this happens, the forwarded content are again not rendered until you refresh.

3. This is more of a sanity check, is it intended that messages from direct chats are allowed to be forwarded to public chats? I was thinking if this is suitable for the privacy of the users.

PaRangger reviewed (dismissed):
Tested on TS2. Everything works great now, thanks for implementing the requested changes! 😄

julian-wls reviewed (dismissed):
Tested on TS6. Works as expected. Re-approve.

anian03 reviewed (changes requested):
Tested on TS6, I noticed a problem with the search bar: It's only possible to enter lowercase letters. So If I try to forward something to a direct chat with "Daniel" I can only find him by typing "aniel" (because "daniel" does not match "Daniel")

<img src="https://github.com/user-attachments/assets/532440dd-ddf4-4d56-b558-92d8f2abe83a" alt="Daniel" width="300">
<img src="https://github.com/user-attachments/assets/6485f555-3f2f-4f0c-b75e-6dc754a90c25" alt="aniel" width="300">

asliayk commented:
> Tested on TS6, I noticed a problem with the search bar: It's only possible to enter lowercase letters. So If I try to forward something to a direct chat with "Daniel" I can only find him by typing "aniel" (because "daniel" does not match "Daniel")
> 
> <img alt="Daniel" width="300" src="https://private-user-images.githubusercontent.com/98647423/397951236-532440dd-ddf4-4d56-b558-92d8f2abe83a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMzA0NDAsIm5iZiI6MTczNTEzMDE0MCwicGF0aCI6Ii85ODY0NzQyMy8zOTc5NTEyMzYtNTMyNDQwZGQtZGRmNC00ZDU2LWI1NTgtOTJkOGYyYWJlODNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI1VDEyMzU0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIyMzlkNDA4Yjk5MTlhNTdmMmVmMWIzOTNiN2YzMWRjMmY5NmI4N2Y2YTJhM2Y0OWUyNTM5Njk5ZTA4ZGY5ZDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.6WFTBjNut-ZmmySD5Mfm0eNAiBXdQwzyzhZs2YhQsBE"> <img alt="aniel" width="300" src="https://private-user-images.githubusercontent.com/98647423/397951237-6485f555-3f2f-4f0c-b75e-6dc754a90c25.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMzA0NDAsIm5iZiI6MTczNTEzMDE0MCwicGF0aCI6Ii85ODY0NzQyMy8zOTc5NTEyMzctNjQ4NWY1NTUtM2YyZi00ZjBjLWI3NWUtNmRjNzU0YTkwYzI1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI1VDEyMzU0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdkM2Y3NzYxZDUyYTcwZTc1ZWI4NDk5NGM5MjU3MzZmYWQ1MTc2YmZmNDMxYjcyYmYwZWYxZWViODUzODNmNWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.oSE8pVfnEmX8TnVfy3kM1LfOROyaJsh4jzrxXmqinpo">

I used the existing logic for user selection. I face this issue while creating a direct conversation as well. So this issue is not related with this PR, but should be solved in another PR.

![image](https://github.com/user-attachments/assets/f379d383-c03b-4211-b2a1-08037aaf4b6d)
![image](https://github.com/user-attachments/assets/a9fa1927-c531-4cf4-81c8-ab809126d4d2)

asliayk commented:
> [Tested on TS2] Functionalities work mostly as expected, except for two quirks while I was playing around.
> 
> 1. After forwarding messages, I logged in to another user and the forwarded content was not rendered at first. After refreshing, they got rendered like normal. This also happened after my 2. point.
>    _directly after logging in_
>    ![image](https://private-user-images.githubusercontent.com/131975714/397528750-f50b1d7b-43a5-4a87-ad23-efe811f91078.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMzA5MzksIm5iZiI6MTczNTEzMDYzOSwicGF0aCI6Ii8xMzE5NzU3MTQvMzk3NTI4NzUwLWY1MGIxZDdiLTQzYTUtNGE4Ny1hZDIzLWVmZTgxMWY5MTA3OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyNVQxMjQzNTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYThmZDQ0ODc4N2IxNWYwNzc1NmIwNTBiNjIyZWM0Njk3N2JjZTE4NDBlNTMyYTBkNTc0OTkyYjFlOTQzZDUwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.fO_Olh3ygEaa5s9aYAx_mr8jLsy7A3x6UYAvlMG2Xmg)
>    _after refreshing_
>    ![image](https://private-user-images.githubusercontent.com/131975714/397528862-e5a1a5c8-80e2-4230-87f4-41f4160a725b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMzA5MzksIm5iZiI6MTczNTEzMDYzOSwicGF0aCI6Ii8xMzE5NzU3MTQvMzk3NTI4ODYyLWU1YTFhNWM4LTgwZTItNDIzMC04N2Y0LTQxZjQxNjBhNzI1Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyNVQxMjQzNTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lYzAyMzk2Mjk1NDc2MzI5MGJiZGI1N2MyYWE0NzE3ZTMzNTljOTUwNTUwMmM4YjA1YzU1ODYxMGEwOGFkNGQyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._aQ3nGChBrVdBFtpPCrDLpvBpopVnCbkJDLzeqJu6qs)
> 2. When forwarding from a private channel to a public channel, the link for "view conversation" appears even if the user is not in the private channel, and clicking on it causes them to access a site they are not supposed to, and reloads them to the all public messages. Is this desired behaviour?
>    ![image](https://private-user-images.githubusercontent.com/131975714/397529193-db8a9268-267b-4833-8950-17b792e6f808.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUxMzA5MzksIm5iZiI6MTczNTEzMDYzOSwicGF0aCI6Ii8xMzE5NzU3MTQvMzk3NTI5MTkzLWRiOGE5MjY4LTI2N2ItNDgzMy04OTUwLTE3Yjc5MmU2ZjgwOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyNVQxMjQzNTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00ODc0NTc2NTBiYzY4MmNhYjBhZDVmMzc2MzU1NmRhZTQ0ZGQyYzRjMzUzMjMwNDliMTg3NGI3MWY0MGZmODliJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Fzl2202G6I4S_PJs1NfWf-Q-Zt3E6fQB70cWo3NQ87U)
>    PS: after this happens, the forwarded content are again not rendered until you refresh.
> 3. This is more of a sanity check, is it intended that messages from direct chats are allowed to be forwarded to public chats? I was thinking if this is suitable for the privacy of the users.

1. I tested it again and didn't encounter this issue. However, some content may load slowly when tested immediately after deployment; I've experienced this in other PRs as well.
2. Since it doesn't crash or show any errors, I think displaying a warning that says "You are not a member of this conversation" is fine for now. In the future, there could be a preview of the channel for users who aren't members, along with a "Join" option at the bottom, similar to Slack.
3. Yes, this is allowed in Slack as well. Artemis already has a confidentiality warning that says, "Keep your personal information safe by keeping it to yourself and not sharing it online" so I think it should be fine.

HawKhiem reviewed (dismissed):
Tested again on TS4. Reapprove

Cathy0123456789 reviewed (approved):
Tested on TS5. Works as described

eylulnc reviewed (approved):
Tested on TS5 works as expected.

krusche commented on src/main/webapp/app/overview/course-conversations/dialogs/forward-message-dialog/forward-message-dialog.component.ts:
```diff
@@ -0,0 +1,238 @@
+export class ForwardMessageDialogComponent implements OnInit, AfterViewInit {
+    channels = signal<ChannelDTO[] | []>([]);
+    users = signal<UserPublicInfoDTO[] | []>([]);
+    postToForward = signal<Post | null>(null);
```
we prefer `undefined` over `null` in the client code, see the client coding guidelines

krusche commented on src/main/webapp/app/overview/course-conversations/dialogs/forward-message-dialog/forward-message-dialog.component.ts:
```diff
@@ -0,0 +1,238 @@
+}
+
+@Component({
+    selector: 'jhi-forward-message-dialog',
```
why is the new component not standalone?

krusche reviewed (changes requested):
1. we prefer `undefined` over `null` in the client code, see the client coding guidelines
2. please make all new components standalone
3. client tests need to pass (including test coverage)

vinceclifford reviewed (dismissed):
Tested on TS5. Works as expected!

asliayk commented:
after resolving merge conflicts and newly added client tests, this PR is ready for review again :)

here are the performance test results:

![image](https://github.com/user-attachments/assets/cedd8c64-9d73-432a-9d6c-7bf2a1ee12d6)

This Locust script tests the forward message workflow under load. Each user:

1. Logs in briefly
2. Fetches messages from a conversation and randomly picks one
3. Creates a new post and forwards the selected message

These steps mimic typical user activity, allowing us to see how the system handles multiple forwarding actions simultaneously. From the attached results, we can observe stable response times and low error rates, indicating that the application can effectively manage concurrent forwarding requests without significant performance issues.

FelberMartin reviewed (changes requested):
Just tested it on TS3, mostly as expected from a technical side. I have two remarks:
1. In dark mode, the forward autocompletion hint row highlighting is quite bright, and makes it hard to read the row that the user currently is hovering over
2.  When searching for users, it looks like we currently only search by loginName. This leads to confusion when I already see the "Test User 12" in the list when typing "test", but then the user disappears again when I type "test user" into the search field. Would it be possible to search users by their full name?

asliayk commented:
> Just tested it on TS3, works mostly as expected from a technical side. I have two remarks:
> 
> 1. In dark mode, the forward autocompletion hint row highlighting is quite bright, and makes it hard to read the row that the user currently is hovering over
> 2. When searching for users, it looks like we currently only search by loginName. This leads to confusion when I already see the "Test User 12" in the list when typing "test", but then the user disappears again when I type "test user" into the search field. Would it be possible to search users by their full name?
> 
>  20250117-1421-36.5808733.mp4

The second issue you mentioned has already been noted. However, I used the preexisting endpoint for user selection, which is also used in the "Create Direct Chat" and "Group Chat" options. In those options, users cannot search by name either; they can only search using the login name. While this is an issue that needs to be addressed, it is not directly related to this PR.

I’m not entirely sure which specific color you are referring to in the first issue. I tried to align it with the existing input sections to make it consistent.

FelberMartin commented:
> The second issue you mentioned has already been noted. However, I used the preexisting endpoint for user selection, which is also used in the "Create Direct Chat" and "Group Chat" options. In those options, users cannot search by name either; they can only search using the login name. While this is an issue that needs to be addressed, it is not directly related to this PR.

Ah I remember having the same problem on the Android app 😅. However I just checked for creating a DM chat, and there searching by username seems to work. Maybe there is already a fix in place?

![image](https://github.com/user-attachments/assets/7b6fc27d-45b6-4712-8e45-b425673da9fb)



--- 
> I’m not entirely sure which specific color you are referring to in the first issue. I tried to align it with the existing input sections to make it consistent.

The color I am refering to, is the bright white of the row item backgrounds, that highlights the row item that the mouse hovers over (see screenshot below). I am not sure what other inputs use this color for highlighting as well, but eg also the selection for the DM chat creation uses a much more pleasant color imo (see screenshot above).

![image](https://github.com/user-attachments/assets/bbde6846-eb82-4849-acfa-7f7f35af4771)

asliayk commented:
> > The second issue you mentioned has already been noted. However, I used the preexisting endpoint for user selection, which is also used in the "Create Direct Chat" and "Group Chat" options. In those options, users cannot search by name either; they can only search using the login name. While this is an issue that needs to be addressed, it is not directly related to this PR.
> 
> Ah I remember having the same problem on the Android app 😅. However I just checked for creating a DM chat, and there searching by username seems to work. Maybe there is already a fix in place?
> 
> ![image](https://private-user-images.githubusercontent.com/45291671/404360995-7b6fc27d-45b6-4712-8e45-b425673da9fb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzcxMzA5OTQsIm5iZiI6MTczNzEzMDY5NCwicGF0aCI6Ii80NTI5MTY3MS80MDQzNjA5OTUtN2I2ZmMyN2QtNDViNi00NzEyLThlNDUtYjQyNTY3M2RhOWZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE3VDE2MTgxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2YTY2YjMyMTYwNWVkODA2YTNlMzhlZTNiYjkxNWVjZjI3MjQ1YWRhYWY5YThmODZiODQxNzc2MzFiMDkyMmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dY9qgyiWy7Qed5d-KW_U5r0SgqZyShQDUfu9TdtnkJY)
> 
> > I’m not entirely sure which specific color you are referring to in the first issue. I tried to align it with the existing input sections to make it consistent.
> 
> The color I am refering to, is the bright white of the row item backgrounds, that highlights the list item that the mouse hovers over (see screenshot below). I am not sure what other inputs use this color for highlighting as well, but eg also the selection for the DM chat creation uses a much more pleasant color imo (see blue highlighted row in screenshot above).
> 
> ![image](https://private-user-images.githubusercontent.com/45291671/404362590-bbde6846-eb82-4849-acfa-7f7f35af4771.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzcxMzA5OTQsIm5iZiI6MTczNzEzMDY5NCwicGF0aCI6Ii80NTI5MTY3MS80MDQzNjI1OTAtYmJkZTY4NDYtZWI4Mi00ODQ5LWFjZmEtN2Y3ZjM1YWY0NzcxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTE3VDE2MTgxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU0YWEyNDlkNzdiNzg3ZjdmOWI3Zjk0ZDAzYzZlMTE0NTQ5M2IxN2RhYmU0NzljMDQ1NjYzZjllZjBjZmQzYTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JfUedDwMhnEz4omrJ2k6YCSKQbwvqOKtjKW9yuNrTCU)

ah I see the color issue now, thank you for noticing this. I think I have solved it now.

However, the user selection does not work for me with usernames on TS3 for direct chat user selection. In my PR, I kept the user selection logic unchanged, as it is in the current implementation. Since the existing logic already has this issue, it also affects this PR. Therefore, I think this issue is unrelated to this PR and should be addressed in a separate PR.


https://github.com/user-attachments/assets/2dafade6-f957-4faa-b768-ee653a35535f

FelberMartin reviewed (dismissed):
The highlight color is much better now, thanks for adjusting!

Regarding the search by username issue, I accidentality tested the user search in the DM dialog on TS2, but you are right, on TS3 that searching by username does not work there.

Good to merge from my side :D

cremertim commented on src/main/webapp/i18n/de/metis.json:
```diff
@@ -165,6 +166,15 @@
+                "addMessage": "Eine Nachricht hinzufügen",
+                "searchBar": "Nach Name oder Kanal suchen",
+                "viewConversation": "Unterhaltung anzeigen",
+                "forwardedFrom": "Forwarded from",
```
Are you intentionally using a english string? Should be  `"forwardedFrom": "Weitergeleitet von",`

asliayk replied:
thanks for notifying this :)

cremertim reviewed (commented):
Code seems good to me, just 1 remark in the translations

HawKhiem reviewed (approved):
Tested on TS3. Reapprove

PaRangger reviewed (approved):
Tested on TS2, works great!

eylulnc reviewed (dismissed):
Tested on TS4 in testing session. It works as described.
One follow up suggestion , deleted message that already been forwarded could be shown as "This message deleted etc"

FelberMartin reviewed (approved):
Tested on TS4, re-approve

anian03 reviewed (dismissed):
Tested on TS4 during testing session, works as described. I agree with @eylulnc that we could consider showing something else when the original message is deleted for a follow up feature

julian-wls reviewed (approved):
Tested on TS4 during testing session. Works as expected, couldn't find any issues.

cremertim reviewed (dismissed):
Latest code changes lgtm. Thanks for adjusting the translation

Cathy0123456789 reviewed (approved):
Retested on TS4. Works without problems :)

zagemello reviewed (dismissed):
Tested on TS1, works as expected!

krusche commented on src/main/webapp/app/shared/metis/posting-reactions-bar/posting-reactions-bar.component.ts:
```diff
@@ -83,6 +89,9 @@ interface ReactionMetaDataMap {
 export class PostingReactionsBarComponent<T extends Posting> implements OnInit, OnChanges {
     protected metisService = inject(MetisService);
     private accountService = inject(AccountService);
+    public conversationService: ConversationService = inject(ConversationService);
```
```suggestion
    private conversationService = inject(ConversationService);
```

krusche commented on src/main/webapp/app/shared/metis/posting-reactions-bar/posting-reactions-bar.component.ts:
```diff
@@ -83,6 +89,9 @@ interface ReactionMetaDataMap {
     protected metisService = inject(MetisService);
     private accountService = inject(AccountService);
+    public conversationService: ConversationService = inject(ConversationService);
+    public modalService: NgbModal = inject(NgbModal);
```
I thinks all injects should be `private` also `metisServce`. In case of usage in test, please us `ts-ignore`

krusche commented on src/main/webapp/app/shared/metis/metis.service.ts:
```diff
@@ -43,6 +47,7 @@ export class MetisService implements OnDestroy {
     protected exerciseService = inject(ExerciseService);
```
can we make this private?

krusche commented on src/main/webapp/app/shared/metis/forwarded-message/forwarded-message.component.ts:
```diff
@@ -0,0 +1,115 @@
+    imports: [ProfilePictureComponent, TranslateDirective, NgClass, FaIconComponent, NgbTooltip, PostingContentComponent, ArtemisTranslatePipe, ArtemisDatePipe],
+})
+export class ForwardedMessageComponent implements AfterViewInit {
+    readonly faShare = faShare;
```
please sort the attributes after type, right now it looks messy

krusche commented on src/main/webapp/app/shared/metis/forwarded-message/forwarded-message.component.ts:
```diff
@@ -0,0 +1,115 @@
+    todayFlag?: string;
+    readonly onNavigateToPost = output<Posting>();
+    messageContent = viewChild<ElementRef>('messageContent');
+    isContentLong: boolean = false;
```
```suggestion
    isContentLong = false;
```

krusche commented on src/main/webapp/app/shared/metis/forwarded-message/forwarded-message.component.ts:
```diff
@@ -0,0 +1,115 @@
+    readonly onNavigateToPost = output<Posting>();
+    messageContent = viewChild<ElementRef>('messageContent');
+    isContentLong: boolean = false;
+    showFullForwardedMessage: boolean = false;
```
```suggestion
    showFullForwardedMessage = false;
```

krusche commented on src/main/webapp/app/shared/metis/forwarded-message.service.ts:
```diff
@@ -0,0 +1,65 @@
+export class ForwardedMessageService {
+    public resourceUrl = 'api/forwarded-messages';
+
+    protected http: HttpClient = inject(HttpClient);
```
```suggestion
    private http = inject(HttpClient);
```

krusche commented on src/main/webapp/app/overview/course-conversations/dialogs/forward-message-dialog/forward-message-dialog.component.ts:
```diff
@@ -0,0 +1,246 @@
+    editorHeight = input<MarkdownEditorHeight>(MarkdownEditorHeight.INLINE);
+    searchInput = viewChild<ElementRef>('searchInput');
+    messageContent = viewChild<ElementRef>('messageContent');
+    public courseManagementService: CourseManagementService = inject(CourseManagementService);
```
```suggestion
    private courseManagementService = inject(CourseManagementService);
```

krusche commented on src/main/webapp/app/overview/course-conversations/dialogs/forward-message-dialog/forward-message-dialog.component.ts:
```diff
@@ -0,0 +1,246 @@
+    showFullForwardedMessage: boolean = false;
+    isContentLong: boolean = false;
+    private cdr = inject(ChangeDetectorRef);
+    public renderer = inject(Renderer2);
```
```suggestion
    private renderer = inject(Renderer2);
```

krusche commented on src/main/webapp/app/overview/course-conversations/dialogs/forward-message-dialog/forward-message-dialog.component.ts:
```diff
@@ -0,0 +1,246 @@
+    isContentLong: boolean = false;
+    private cdr = inject(ChangeDetectorRef);
+    public renderer = inject(Renderer2);
+    public activeModal = inject(NgbActiveModal);
```
```suggestion
    private activeModal = inject(NgbActiveModal);
```

asliayk replied:
since `activeModal` is used in another file, I converted it to protected

krusche commented on src/main/webapp/app/overview/course-conversations/dialogs/forward-message-dialog/forward-message-dialog.component.ts:
```diff
@@ -0,0 +1,246 @@
+    providers: [MetisService, LinkPreviewService, LinkifyService],
+})
+export class ForwardMessageDialogComponent implements OnInit, AfterViewInit {
+    channels = signal<ChannelDTO[] | []>([]);
```
please sort the attributes after type and leave empty lines

krusche commented on src/main/webapp/app/overview/course-conversations/course-conversations.module.ts:
```diff
@@ -33,6 +33,7 @@ import { ProfilePictureComponent } from 'app/shared/profile-picture/profile-pict
 import { SavedPostsComponent } from 'app/overview/course-conversations/saved-posts/saved-posts.component';
 import { PostingSummaryComponent } from 'app/overview/course-conversations/posting-summary/posting-summary.component';
 import { InfiniteScrollDirective } from 'ngx-infinite-scroll';
+import { ForwardMessageDialogComponent } from 'app/overview/course-conversations/dialogs/forward-message-dialog/forward-message-dialog.component';
```
We actually want to get rid of Angular client modules. Would it make sense to remove this one here or in another PR?

asliayk replied:
I removed `ForwardMessageDialogComponent` from this module. I think additional changes can be made in follow-up PRs.

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/repository/ForwardedMessageRepository.java:
```diff
@@ -0,0 +1,25 @@
+@Repository
+public interface ForwardedMessageRepository extends ArtemisJpaRepository<ForwardedMessage, Long> {
+
+    @Query("SELECT fm FROM ForwardedMessage fm WHERE fm.destinationPost.id IN :destinationPostIds")
```
can you write this into multiple lines to improve readability?

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/repository/ForwardedMessageRepository.java:
```diff
@@ -0,0 +1,25 @@
+    @Query("SELECT fm FROM ForwardedMessage fm WHERE fm.destinationPost.id IN :destinationPostIds")
+    Set<ForwardedMessage> findAllByDestinationPostIds(@Param("destinationPostIds") Set<Long> destinationPostIds);
+
+    @Query("SELECT fm FROM ForwardedMessage fm WHERE fm.destinationAnswerPost.id IN :destinationAnswerPostIds")
```
can you write this into multiple lines to improve readability?

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/AnswerMessageResource.java:
```diff
@@ -92,4 +95,11 @@ public ResponseEntity<Void> deleteAnswerMessage(@PathVariable Long courseId, @Pa
         return ResponseEntity.ok().build();
     }
+
+    @GetMapping("courses/{courseId}/answer-messages/source-answer-posts")
```
```suggestion
    @GetMapping("communication/courses/{courseId}/answer-messages-source-posts")
```

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/AnswerMessageResource.java:
```diff
@@ -92,4 +95,11 @@ public ResponseEntity<Void> deleteAnswerMessage(@PathVariable Long courseId, @Pa
+    @GetMapping("courses/{courseId}/answer-messages/source-answer-posts")
+    @EnforceAtLeastStudent
+    public ResponseEntity<List<AnswerPost>> getSourceAnswerPostsByIds(@PathVariable Long courseId, @RequestParam List<Long> answerPostIds) {
+        List<AnswerPost> answerPosts = answerMessageService.findByIdIn(answerPostIds);
```
you are missing a security check that the user is at least student in the course, either change the annoation above or add an additional check here. This should be part of the resource and not be done in a service (except it's clear from the method names)

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ConversationMessageResource.java:
```diff
@@ -203,4 +203,19 @@ public ResponseEntity<Post> updateDisplayPriority(@PathVariable Long courseId, @
+    @GetMapping("courses/{courseId}/messages/source-posts")
+    @EnforceAtLeastStudent
+    public ResponseEntity<List<Post>> getSourcePostsByIds(@PathVariable Long courseId, @RequestParam List<Long> postIds) {
+        List<Post> posts = conversationMessagingService.getMessageByIds(postIds);
```
you are missing a security check that the user is at least student in the course, either change the annoation above or add an additional check here. This should be part of the resource and not be done in a service (except it's clear from the method names)

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ConversationMessageResource.java:
```diff
@@ -203,4 +203,19 @@ public ResponseEntity<Post> updateDisplayPriority(@PathVariable Long courseId, @
+     * @return ResponseEntity with status 200 (OK) containing the list of posts in the response body,
+     *         or with status 400 (Bad Request) if the checks on user, course or post validity fail
+     */
+    @GetMapping("courses/{courseId}/messages/source-posts")
```
```suggestion
    @GetMapping("communication/courses/{courseId}/messages-source-posts")
```

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ForwardedMessageResource.java:
```diff
@@ -0,0 +1,104 @@
+ */
+@Profile(PROFILE_CORE)
+@RestController
+@RequestMapping(value = "api/")
```
```suggestion
@RequestMapping("api/communication/")
```

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ForwardedMessageResource.java:
```diff
@@ -0,0 +1,104 @@
+     */
+    @PostMapping("forwarded-messages")
+    @EnforceAtLeastStudent
+    public ResponseEntity<ForwardedMessageDTO> createForwardedMessage(@RequestBody ForwardedMessage forwardedMessage) throws URISyntaxException {
```
please use a DTO here, we want to avoid using entity objects completely in the REST layer in the future

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ForwardedMessageResource.java:
```diff
@@ -0,0 +1,104 @@
+     */
+    @GetMapping("forwarded-messages")
+    @EnforceAtLeastStudent
+    public ResponseEntity<List<ForwardedMessagesGroupDTO>> getForwardedMessages(@RequestParam Set<Long> ids, @RequestParam String type) {
```
What kind of `ids` do you send here? please use a specific name, we want to avoid `id` and `ids` as this is too generic

krusche replied:
e.g. `postingIds` could make more sense here

krusche commented on src/main/java/de/tum/cit/aet/artemis/communication/web/ForwardedMessageResource.java:
```diff
@@ -0,0 +1,104 @@
+    @EnforceAtLeastStudent
+    public ResponseEntity<List<ForwardedMessagesGroupDTO>> getForwardedMessages(@RequestParam Set<Long> ids, @RequestParam String type) {
+
+        log.debug("GET getForwardedMessages invoked with ids {} and type {}", ids, type);
```
looks like security checks (student in course) and student has access to those forwarded messages is missing

krusche reviewed (changes requested):
Please add security checks (important!!!) and adapt the REST calls as specified in my inline comments. The requirement with the module prefix is new, but I'd like to address this here as well. However, the missing security checks are really problematic, please take this into account in general (we have documentation in the server coding guidelines)
